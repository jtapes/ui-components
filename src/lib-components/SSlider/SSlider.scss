@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';
@import '~vue-slider-component/dist-css/vue-slider-component.css';

@mixin triangle($size, $color, $direction) {
  width: 0;
  height: 0;
  @if ($direction == top) or
    ($direction == bottom) or
    ($direction == right) or
    ($direction == left)
  {
    border-style: solid;
    border-width: $size / 2;
    border-color: transparent;
    @if $direction == top {
      border-bottom-color: $color;
    } @else if $direction == right {
      border-left-color: $color;
    } @else if $direction == bottom {
      border-top-color: $color;
    } @else if $direction == left {
      border-right-color: $color;
    }
  }
}

@mixin arrow($size, $color) {
  &::after {
    content: '';
    position: absolute;
  }

  &-top {
    &::after {
      @include triangle($size, $color, bottom);

      left: 50%;
      top: 100%;
      transform: translate(-50%, 0);
    }
  }

  &-bottom {
    &::after {
      @include triangle($size, $color, top);

      left: 50%;
      bottom: 100%;
      transform: translate(-50%, 0);
    }
  }

  &-left {
    &::after {
      @include triangle($size, $color, right);

      left: 100%;
      top: 50%;
      transform: translate(0, -50%);
    }
  }

  &-right {
    &::after {
      @include triangle($size, $color, left);

      right: 100%;
      top: 50%;
      transform: translate(0, -50%);
    }
  }
}

$themeColor: $color--primary-accent !default;
$disabledOpacity: 0.5 !default;

$bgColor: #ccc !default;
$railBorderRadius: 15px !default;

$dotShadow: 0.5px 0.5px 2px 1px rgba(0, 0, 0, 0.32) !default;
$dotShadowFocus: 0 0 1px 2px rgba($themeColor, 0.36) !default;
$dotBgColor: #fff !default;
$dotBgColorDisable: #ccc !default;
$dotBorderRadius: 50% !default;

$tooltipBgColor: $themeColor !default;
$tooltipColor: #fff !default;
$tooltipBorderRadius: 5px !default;
$tooltipPadding: 2px 5px !default;
$tooltipMinWidth: 20px !default;
$tooltipArrow: 10px !default;
$tooltipFontSize: 14px !default;

$stepBorderRadius: 50% !default;
$stepBgColor: rgba(0, 0, 0, 0.16) !default;

$labelFontSize: 14px !default;

/* component style */
.vue-slider-disabled {
  opacity: $disabledOpacity;
  cursor: not-allowed;
}

/* rail style */
.vue-slider-rail {
  border-radius: $railBorderRadius;
  background-color: $bgColor;
}

/* process style */
.vue-slider-process {
  border-radius: $railBorderRadius;
  background-color: $themeColor;
}

/* mark style */
.vue-slider-mark {
  z-index: 4;

  &:first-child,
  &:last-child {
    .vue-slider-mark-step {
      display: none;
    }
  }

  @at-root &-step {
    border-radius: $stepBorderRadius;
    width: 100%;
    height: 100%;
    background-color: $stepBgColor;
  }

  @at-root &-label {
    font-size: $labelFontSize;
    white-space: nowrap;
  }
}

/* dot style */
.vue-slider-dot {
  width: $size--16 !important; /* stylelint-disable-line declaration-no-important */
  height: $size--16 !important; /* stylelint-disable-line declaration-no-important */

  @at-root &-handle {
    border: 1px solid $themeColor;
    border-radius: $dotBorderRadius;
    width: 100%;
    height: 100%;
    box-shadow: $dotShadow;
    box-sizing: border-box;
    background-color: $dotBgColor;
    cursor: pointer;

    @at-root &-focus {
      box-shadow: $dotShadowFocus;
    }
    @at-root &-disabled {
      background-color: $dotBgColorDisable;
      cursor: not-allowed;
    }
  }

  @at-root &-tooltip {
    @at-root &-inner {
      @include arrow($tooltipArrow, inherit);

      border-color: $tooltipBgColor;
      border-radius: $tooltipBorderRadius;
      padding: $tooltipPadding;
      min-width: $tooltipMinWidth;
      box-sizing: content-box;
      font-size: $tooltipFontSize;
      white-space: nowrap;
      text-align: center;
      color: $tooltipColor;
      background-color: $tooltipBgColor;
    }
  }

  @at-root &-tooltip-wrapper {
    opacity: 0;
    transition: all 0.3s;
    @at-root &-show {
      opacity: 1;
    }
  }
}
