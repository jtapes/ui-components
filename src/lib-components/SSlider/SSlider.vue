<template>
  <vue-slider
    :value="value"
    v-bind="$attrs"
    @change="change"
    @drag-end="onDragEnd"
    @drag-start="onDragStart"
  />
</template>
<script lang="ts">
import { Component, Model, Vue } from 'vue-property-decorator'

import VueSlider from 'vue-slider-component/dist-css/vue-slider-component.umd.min.js'

@Component({ components: { VueSlider } })
export default class SSlider extends Vue {
  @Model('change')
  value!: [number, number] | number

  change(value: [number, number] | number): void {
    this.$emit('change', value)
  }

  onDragEnd(): void {
    this.$emit('drag-end')
  }

  onDragStart(): void {
    this.$emit('drag-start')
  }
}
</script>
