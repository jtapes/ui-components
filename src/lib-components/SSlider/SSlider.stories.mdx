import {action} from '@storybook/addon-actions'
import {config, withDesign} from 'storybook-addon-designs'
import {withTests} from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Canvas,
} from '@storybook/addon-docs/blocks'
import results from '/.jest-test-results.json';
import SSlider from '@/lib-components/SSlider/SSlider.vue';

<Meta
  title="DESIGN SYSTEM/Atom/Slider"
  component={SSlider}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    value: {
      control: 'array',
      defaultValue: [10, 30],
      description: 'Положение контроллера',
    },
    max: {
      control: 'number',
      defaultValue: 100,
      description: 'Максимальная длина',
      type: 'Number'
    },
    min: {
      control: 'number',
      defaultValue: 0,
      description: 'Минимальная длина',
      type: 'Number'
    },
    tooltip: {
      control: 'text',
      defaultValue: 'none',
      description: 'Поведение маркеров',
      type: 'String',
    },
    enableCross: {
      control: 'boolean',
      description: 'Перемещение через контроллеры',
      defaultValue: false,
      type: 'Boolean'
    },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-Market-Catalog-%E2%80%A2-Product-Page?node-id=1902%3A1902',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

export const Template = (args, {argTypes}) => ({
  props: Object.keys(argTypes),
  components: {SSlider},
  template: `<s-slider style="max-width: 200px;" class="mt-24" :value="currentValue" @change="change" :tooltip="tooltip" :max="max" :min="min" :enable-cross="enableCross" />`,
  data() {
    return {
      currentValue: this.value
    }
  },
  watch: {
    value(v) {
      this.currentValue = v
    }
  },
  methods: {
    change(value) {
      action('change')(value)
      this.currentValue = value
    },
  },
})

<Canvas withToolbar withSource="none">
  <Story name="Range Slider">
    {Template.bind({})}
  </Story>
  <Story name="Slider" args={{ value: 30, tooltip: 'active' }}>
    {Template.bind({})}
  </Story>
</Canvas>


<Description of={SSlider} />
<ArgsTable of={SSlider} />
<Source code={`<s-slider :value="[10, 30]" />`} />

