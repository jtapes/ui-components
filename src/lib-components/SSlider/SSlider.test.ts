import { mount } from '@vue/test-utils'
import SSlider from '@/lib-components/SSlider/SSlider.vue'

describe('SSlider', () => {
  const wrapper = mount(SSlider, {
    propsData: {
      value: 150,
      max: 200,
      min: 100,
    },
  })
  test('process', () => {
    const position = 50
    const reg = RegExp(`width: ${position}%`)
    const styles = wrapper.find('.vue-slider-process').attributes('style') || ''
    expect(reg.test(styles)).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('range', async () => {
    await wrapper.setProps({ value: [110, 150] })

    const position = 40
    const leftPosition = 10
    const reg = RegExp(`left: ${leftPosition}%; width: ${position}%;`)
    const styles = wrapper.find('.vue-slider-process').attributes('style') || ''
    expect(reg.test(styles)).toBeTruthy()
    expect(wrapper.element).toMatchSnapshot()
  })
})
