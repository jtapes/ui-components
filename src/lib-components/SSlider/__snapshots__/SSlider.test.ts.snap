// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSlider process 1`] = `
<div
  class="vue-slider vue-slider-ltr"
  max="200"
  min="100"
  style="padding: 7px 0px; width: auto; height: 4px;"
>
  <div
    class="vue-slider-rail"
  >
    <div
      class="vue-slider-process"
      style="height: 100%; top: 0px; left: 0%; width: 50%; transition-property: width,left; transition-duration: 0.5s;"
    />
    <div
      aria-orientation="horizontal"
      aria-valuemax="200"
      aria-valuemin="100"
      aria-valuenow="150"
      aria-valuetext="150"
      class="vue-slider-dot vue-slider-dot-hover"
      role="slider"
      style="width: 14px; height: 14px; transform: translate(-50%, -50%); top: 50%; left: 50%; transition: left 0.5s;"
      tabindex="0"
    >
      <div
        class="vue-slider-dot-handle"
      />
      <div
        class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top"
      >
        <div
          class="vue-slider-dot-tooltip-inner vue-slider-dot-tooltip-inner-top"
        >
          <span
            class="vue-slider-dot-tooltip-text"
          >
            150
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SSlider range 1`] = `
<div
  class="vue-slider vue-slider-ltr"
  max="200"
  min="100"
  style="padding: 7px 0px; width: auto; height: 4px;"
>
  <div
    class="vue-slider-rail"
  >
    <div
      class="vue-slider-process"
      style="height: 100%; top: 0px; left: 10%; width: 40%; transition-property: width,left; transition-duration: 0.5s;"
    />
    <div
      aria-orientation="horizontal"
      aria-valuemax="200"
      aria-valuemin="100"
      aria-valuenow="110"
      aria-valuetext="110"
      class="vue-slider-dot vue-slider-dot-hover"
      role="slider"
      style="width: 14px; height: 14px; transform: translate(-50%, -50%); top: 50%; left: 10%; transition: left 0.5s;"
      tabindex="0"
    >
      <div
        class="vue-slider-dot-handle"
      />
      <div
        class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top"
      >
        <div
          class="vue-slider-dot-tooltip-inner vue-slider-dot-tooltip-inner-top"
        >
          <span
            class="vue-slider-dot-tooltip-text"
          >
            110
          </span>
        </div>
      </div>
    </div>
    <div
      aria-orientation="horizontal"
      aria-valuemax="200"
      aria-valuemin="100"
      aria-valuenow="150"
      aria-valuetext="150"
      class="vue-slider-dot vue-slider-dot-hover"
      role="slider"
      style="width: 14px; height: 14px; transform: translate(-50%, -50%); top: 50%; left: 50%; transition: left 0.5s;"
      tabindex="0"
    >
      <div
        class="vue-slider-dot-handle"
      />
      <div
        class="vue-slider-dot-tooltip vue-slider-dot-tooltip-top"
      >
        <div
          class="vue-slider-dot-tooltip-inner vue-slider-dot-tooltip-inner-top"
        >
          <span
            class="vue-slider-dot-tooltip-text"
          >
            150
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
